source/auth/auth.rst

```diff
@@ -1251,172 +1212,234 @@ mechanism
-PROVIDER_NAME is set but credentials cannot be automatically obtained,
-the driver MUST raise an error.
+Conversation
+````````````
+If using the ``PrincipalStepRequest`` first, the conversation will look like:
 
+| C: :javascript:`{saslStart: 1, mechanism: "MONGODB-OIDC", payload: BinData(0, "...")}`
+| S: :javascript:`{conversationId : 1, payload: BinData(0,"..."), done: false, ok: 1}`
+| C: :javascript:`{saslContinue: 1, conversationId: 1, payload: BinData(0, "...")}`
+| S: :javascript:`{conversationId: 1, payload: BinData(0,".."), done: true, ok: 1}`
 
-Supported Service Providers
-```````````````````````````
+If using the ``JwtStepRequest`` directly, the conversation will look like:
 
-Drivers MUST support obtaining credentials for a service for "aws", given
-by the PROVIDER_NAME mechanism property.  In all cases the acquired token
-will be given as the ``jwt`` argument and the JwtStepRequest MUST be made immediately, as part of speculative authentication if appropriate, skipping the PrincipalStepRequest.
-Drivers MUST raise an error if both a PROVIDER_NAME and username are
-given, since using a service will not use the username.
+| C: :javascript:`{saslStart: 1, mechanism: "MONGODB-OIDC", payload: BinData(0, "...")}`
+| S: :javascript:`{conversationId : 1, payload: BinData(0,"..."), done: true, ok: 1}`
 
-AWS
-___
+Drivers MUST NOT send a ``PrincipalStepRequest`` when performing machine authentication
+or when there is a cached Access Token.  Drivers MUST instead  use ``saslStart`` with a ``JwtStepRequest``.
 
-When the PROVIDER_NAME mechanism property is set to "aws", the driver MUST
-attempt to read the value given by the ``AWS_WEB_IDENTITY_TOKEN_FILE`` and
-interpret it as a file path.  The contents of the file are read as the
-access token.  If the path does not exist or cannot be read, or the environment variable does not exist, the driver MUST raise an error.
+Caching
+```````
+Drivers MUST cache the ``IdPInfo``, Access Token, and Refresh Token associated with each
+``MongoClient``.  If any operation fails the driver MUST clear the Access Token.  The Refresh Token is handled differently,
+see the ``Reauthentication`` section below.
 
+Drivers MUST also use a token generation id that is incremented whenever a new Access Token is retrieved.
+When a connection succeeds, the token generation id MUST be associated with or stored on the connection object.
+This value is used in the ``Reauthentication`` section below.
 
-Caching Credentials
-```````````````````
+Speculative Authentication
+``````````````````````````
+Drivers MUST implement speculative authentication for MONGODB-OIDC during the ``hello`` handshake.
+If there is a cached Access Token the ``JwtStepRequest`` SASL command will be used as the speculation command.
+If there is no cached Access Token, the ``PrincipalStepRequest`` will be used as the speculation command.
+The driver MUST NOT call the callback during speculative authentication.
```

Comment:
I think we want two different callbacks, but I still think we shouldn't call it, because for Azure and perhaps other "custom" types there is an http request.